### DISCUSSION TITLE
Auto-wiring parameters without factory-callback

### DISCUSSION THREAD

**Discussion by jkrzefski** [clue/framework-x#192] [Ideas]
I recently made a [pull-request to allow auto-wiring of container-parameters without using the factory-callback method](https://github.com/clue/framework-x/pull/187). Simply put, that would allow you to turn this:

```php
$container = new FrameworkX\Container([
    UserController::class => function (string $hostname): UserController {
        return new UserController($hostname);
    },
    'hostname' => 'Some string value',
]);
```

into this:

```php
$container = new FrameworkX\Container([
    'hostname' => 'Some string value',
]);
```

This is accomplished by accepting string parameters as arguments in auto-wiring for services.

But as @SimonFrings mentioned, this feature might have unwanted side-effects. I am not able to grasp the entire impact of this change, so that's what this discussion is for. Maybe there can be name-collisions. Maybe some other issues may come up. Please let's discuss this here to find a conclusion and hopefully make everything easier for everyone.

**Comment by SimonFrings** [comment]
Yes, as @jkrzefski said we have to take a deeper look into the consequences. One thing I know for sure is, that if we implement this we have to do it right so we don't have to revert it at a later time (which should never be the answer). I spoke with @clue about this and he mentioned some conflicts that could come up, so I'll use this opportunity to ping him so he can go more into detail 😅

**Reply by clue** [reply]
The above suggestion sounds completely reasonable, but I'm afraid this might lead to unforeseen consequences if we blindly pass the parameter down the dependency tree automatically. For instance, in your example `$hostname` obviously refers to the constructor argument in your `UserController`, but what happens if another `IndexController` requires a `MysqlClient` which happens to use the same variable name?

I definitely like the idea of making injecting variables easier and require less boilerplate, but I wonder what the container configuration format could look like.

At the moment, this is not an issue as you have to manually write some boilerplate to explicitly inject the correct variables.

**Comment by skydiablo** [comment]
there is already a really powerful container/DI framework, why we do not use this? => https://github.com/symfony/dependency-injection

**Reply by SimonFrings** [reply]
@skydiablo I would say go for it if you want to use it, I won't stop you  😄
